process/block/alteredaccounts/alteredAccountsProvider_test.go

```diff
@@ -0,0 +1,430 @@
+			},
+			"hash1": &transaction.Transaction{
+				SndAddr: []byte("sender shard - tx1"),
+				RcvAddr: []byte("receiver shard - tx1"),
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, 2, len(res))
+
+	for key := range res {
+		decodedKey, _ := args.AddressConverter.Decode(key)
+		require.True(t, strings.HasPrefix(string(decodedKey), "sender"))
+	}
+}
+
+func testExtractAlteredAccountsFromPoolReceiverShard(t *testing.T) {
+	t.Parallel()
+
+	args := getMockArgs()
+	args.ShardCoordinator = &testscommon.ShardsCoordinatorMock{
+		ComputeIdCalled: func(address []byte) uint32 {
+			if strings.HasPrefix(string(address), "sender") {
+				return 0
+			}
+
+			return 1
+		},
+		SelfIDCalled: func() uint32 {
+			return 1
+		},
+	}
+	aap, _ := NewAlteredAccountsProvider(args)
+
+	res, err := aap.ExtractAlteredAccountsFromPool(&indexer.Pool{
+		Txs: map[string]data.TransactionHandler{
+			"hash0": &transaction.Transaction{
+				SndAddr: []byte("sender shard - tx0"),
+				RcvAddr: []byte("receiver shard - tx0"),
+			},
+			"hash1": &transaction.Transaction{
+				SndAddr: []byte("sender shard - tx1"),
+				RcvAddr: []byte("receiver shard - tx1"),
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, 2, len(res))
+
+	for key := range res {
+		decodedKey, _ := args.AddressConverter.Decode(key)
+		require.True(t, strings.HasPrefix(string(decodedKey), "receiver"))
+	}
+}
+
+func testExtractAlteredAccountsFromPoolBothSenderAndReceiverShards(t *testing.T) {
+	t.Parallel()
+
+	args := getMockArgs()
+	args.ShardCoordinator = &testscommon.ShardsCoordinatorMock{
+		ComputeIdCalled: func(address []byte) uint32 {
+			shardIdChar := string(address)[5]
+
+			return uint32(shardIdChar - '0')
+		},
+		SelfIDCalled: func() uint32 {
+			return 0
+		},
+	}
+	aap, _ := NewAlteredAccountsProvider(args)
+
+	res, err := aap.ExtractAlteredAccountsFromPool(&indexer.Pool{
+		Txs: map[string]data.TransactionHandler{
+			"hash0": &transaction.Transaction{ // intra-shard 0, different addresses
+				SndAddr: []byte("shard0 addr - tx0"),
+				RcvAddr: []byte("shard0 addr 2 - tx0"),
+			},
+			"hash1": &transaction.Transaction{ // intra-shard 0, same addresses
+				SndAddr: []byte("shard0 addr 3 - tx1"),
+				RcvAddr: []byte("shard0 addr 3 - tx1"),
+			},
+			"hash2": &transaction.Transaction{ // cross-shard, sender in shard 0
+				SndAddr: []byte("shard0 addr - tx2"),
+				RcvAddr: []byte("shard1 - tx2"),
+			},
+			"hash3": &transaction.Transaction{ // cross-shard, receiver in shard 0
+				SndAddr: []byte("shard1 addr - tx3"),
+				RcvAddr: []byte("shard0 addr - tx3"),
+			},
+			"hash4": &transaction.Transaction{ // cross-shard, no address in shard 0
+				SndAddr: []byte("shard2 addr - tx4"),
+				RcvAddr: []byte("shard2 addr - tx3"),
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Equal(t, 5, len(res))
+
+	for key := range res {
+		decodedKey, _ := args.AddressConverter.Decode(key)
+		require.True(t, strings.HasPrefix(string(decodedKey), "shard0"))
```

Comment:
I highly recommend here and in the above tests that you check for the full address instead of prefix only.
E.g.: Here you have 6 addresses which have `shard0` prefix, and you expect 5 of them as output. Which one of them do you expect? How can you be sure that those 5 are the ones you actually expect? The prefix is not enough.

It would be also useful to check if for the corresponding address, you have loaded the correct balance/nonce

Reply:
no need for full address check because the output is a map and keys are unique. balance/nonce checks are done on other unit tests

Reply:
The only reason why this particular test works without checking for full address is because at `hash1` you have a `self` transaction, address being: `shard0 addr 3 - tx1`. But again, looking at your function as a `black-box`,  you cannot know how addresses are stored in the map. You could even think that the addresses are altered during execution. Even though it holds unique keys, you still don't know exactly which keys are stored (again, looking at it as a black-box).

Reply:
ok, check that the resulting map contains all the addresses I am interested in